src/amber/router/file.cr

```diff
@@ -0,0 +1,26 @@
+require "tempfile"
+
+module Amber::Router
+  struct File
+    getter file : Tempfile
+    getter filename : String?
+    getter headers : HTTP::Headers
+    getter creation_time : Time?
+    getter modification_time : Time?
+    getter read_time : Time?
+    getter size : UInt64?
+
+    def initialize(upload)
+      @filename = upload.filename
```

Comment:
Is `upload.filename` direct from end user form submission? Is naming a file on the disk by a string provided from some unknown source a dangerous idea? Could this be changed so that the tempfile has a uuid name, but the `Router::File#filename` method still returns the name that was sent in the request?

Reply:
@robacarp @eliasjpr  Right, it would be nice to use a tempfile because what if the original filename is maliciously reported as e.g. "../../../../../tmp/myfile"? It would allow one to save it to an unintended directory.

But more importantly - why is a temp file being saved at all, instead of the file contents persisting in memory until the user decides to save them somewhere (or not save them at all)?

Reply:
The reason to write the file to disk is to offload the server memory. Think what would happened if your crystal app is running on 1gb of ram and and you upload a 2gb file? That will use all the ram available to the server. Also the temp directory is the default behavior for most web servers I do see some good reasons why rename the uploaded file on server

Reply:
@eliasjpr yes, but the file contents already are in memory. The data transfer of file contents won't begin when you save the file, but it has already been completed and the whole contents are in memory. Or I am wrong on that?

Reply:
I don't think keeping file in memory is advisable or good practice. Reasons that I see is:

- One form can have multiple files to upload
- File uploads can be streamed (Chunked uploads)
- https://github.com/rack/rack/blob/master/lib/rack/multipart/uploaded_file.rb#L14

Reply:
Writing file to tempfile is pretty standard. In regard to the name of the file Tempfile already handles adding a random string for the filename. https://crystal-lang.org/api/0.23.0/Tempfile.html

In terms of storing, upload.body in memory, I don't think that would be very useful, and I'd be afraid of memory usage increasing out of control as unclosed connections built up. As it currently stands the request has this already but storing the parsed multipart file in memory would over double the memory per request.... WDYGT?

Reply:
If tempfile is already randomizing, we should just not use the file name at all. It’s not worth the potential risk.

Reply:
This isn't actually even a new change it just shows that way because @eliasjpr moved a bunch of files around and renamed them. It's useful to store the filename so that when doing something with it in the controller we know what it was originally called.